src/http/one/TeChunkedParser.cc

```diff
@@ -132,13 +137,22 @@ Http::One::TeChunkedParser::parseChunkMetadataSuffix(Tokenizer &tok)
     // other exceptions bubble up to kill message parsing
 }
 
-/// Parses the chunk-ext list (RFC 7230 section 4.1.1 and its Errata #4667):
+/// Parses the chunk-ext list (RFC 9112 section 7.1.1:
 /// chunk-ext = *( BWS ";" BWS chunk-ext-name [ BWS "=" BWS chunk-ext-val ] )
 void
 Http::One::TeChunkedParser::parseChunkExtensions(Tokenizer &tok)
 {
     do {
-        ParseBws(tok); // Bug 4492: IBM_HTTP_Server sends SP after chunk-size
+        // Bug 4492: IBM_HTTP_Server sends SP after chunk-size
+        // causing RFC 7230 Errata #4667
+
+        // Use strict BWS definition here to avoid consuming
+        // from our callers "[chunk-ext] CRLF" sequence when
+        // using the relaxed_header_parser.
+        if (const auto count = tok.skipAll(CharacterSet::WSP)) {
+            // Generating BWS is a MUST-level violation so warn about it as needed.
+            debugs(33, ErrorLevel(), "found " << count << " BWS octets");
+        }
 
```

Comment:
Please do change portions of the existing grammar/code "to avoid consuming from our callers". The latest PR attempt (i.e. PR commit 4a39435cf8568fabae2bbb8c84d753569bf04918) is buggy AFAICT -- it does not cover some cases -- but it is the whole _approach_ that is flawed. We should change the approach rather than chase and fix individual PR bugs (and then wait for similar bugs to appear in future code).

Official Http::One::TeChunkedParser::parseChunkExtensions() is buggy: It promises to parse `chunk-ext` (as defined by the corresponding ABNF), but it also parses other sequences (e.g., `BWS`). This PR should fix _that_ bug. It currently does not (it replaces some `BWS` with other sequences and increases code complexity/duplication, but the underlying bug remains unaddressed).

I suggest renaming parseChunkExtensions() parameter from tok to callerTok and updating callerTok at the existing commit point (at the end of each loop iteration). All current PR changes inside this method should be removed. Here is how the fixed version of the method loop body would look like (GitHub does not allow me to post this as a suggestion):

```C++
        auto tok = callerTok;

        ParseBws(tok); // Bug 4492: IBM_HTTP_Server sends SP after chunk-size

        if (!tok.skip(';'))
            return; // reached the end of extensions (if any)

        parseOneChunkExtension(tok);
        buf_ = tok.remaining(); // got one extension
        callerTok = tok;
```

In the PR diff, you should see no loop changes except three new lines: two at the top (including an empty line) and one at the bottom of the loop body code.

Http::One::TeChunkedParser::parseOneChunkExtension() has a very similar bug and needs a very similar fix (instead of the current PR changes).

I did not check other TeChunkedParser methods. Please do so.

Reply:
FTR; the issue currently is that "relaxed" `ParseBws()` consumes characters outside the [`BWS` definition](https://www.rfc-editor.org/rfc/rfc9110.html#section-5.6.3).
The software violating the RFC sender requirements is known to use `SP` so we can be strict with these syntax tolerances. Remember the goal behind this PR is to increase standard compliance, not reduce it.

Reply:
> FTR; the issue currently is that "relaxed" `ParseBws()` consumes characters outside the [`BWS` definition](https://www.rfc-editor.org/rfc/rfc9110.html#section-5.6.3).

While there are many "current issues" that can be discussed (in their proper contexts), _this_ change request is about TeChunkedParser problems that are _not_ tied to any specific ParseBws() implementation. Addressing this change request does not imply, does not require, and cannot be done by changing ParseBws() (or replacing ParseBws() calls with ad hoc code that follows RFC definition of BWS).

> Remember the goal behind this PR is to increase standard compliance, not reduce it.

Addressing this change request will fix bugs without reducing standard compliance.

I recommend carefully rereading the change request with the intent to understand the actual problem it describes and the solution it outlines. Neither are related to the first response to this change request (partially quoted above).

Reply:
Done by [8caa384](https://github.com/squid-cache/squid/pull/1498/commits/8caa3843216b26b6028ae0b40e83acc34db2583b)